src/main/java/se/liu/ida/hefquin/engine/data/mappings/impl/EntityMappingImpl.java

```diff
@@ -25,20 +27,86 @@ public EntityMappingImpl( final Graph mappingDescription ) {
 	}
 
 	protected void parseMappingDescription( final Graph mappingDescription ) {
-		// TODO: populate both g2lMap and l2gMap based on the owl:sameAs
+		// populate both g2lMap and l2gMap based on the owl:sameAs
 		// statements in the given RDF graph; assume the the subject of
 		// any owl:sameAs statement is a global URI and the object is the
 		// corresponding local URI
 		final Iterator<Triple> it = mappingDescription.find(Node.ANY, OWL2.sameAs.asNode(), Node.ANY);
 		while ( it.hasNext() ) {
-			// TODO ...
+			if(g2lMap.containsKey(it.next().getSubject())) {
+				// There is already at least one local URI for this global URI,
+				// We don't want to override it, so we'll add the new one to the set.
+				final Set<Node> locals = g2lMap.get(it.next().getSubject());
+				locals.add(it.next().getObject());
+				g2lMap.replace(it.next().getSubject(),locals);
+			} else {
+				// g2lmap doesn't contain any local URI for this global yet
+				final Set<Node> locals = new HashSet<Node>();
+				locals.add(it.next().getObject());
+				g2lMap.put(it.next().getSubject(),locals);
+			}
+			
+			// The same for l2g.
+			if(l2gMap.containsKey(it.next().getObject())) {
+				final Set<Node> globals = l2gMap.get(it.next().getObject());
+				globals.add(it.next().getSubject());
+				l2gMap.replace(it.next().getObject(),globals);
+			} else {
+				final Set<Node> globals = new HashSet<Node>();
+				globals.add(it.next().getSubject());
+				l2gMap.put(it.next().getObject(),globals);
+			}
 		}
 	}
 
 	@Override
 	public Set<TriplePattern> applyToTriplePattern( final TriplePattern tp ) {
-		// TODO: implement this function
-		return Collections.singleton(tp);
+		final Set<Triple> workingSet = new HashSet<Triple>();
+		if(g2lMap.containsKey(tp.asJenaTriple().getSubject())) {
+			final Iterator<Node> it = g2lMap.get(tp.asJenaTriple().getSubject()).iterator();
```

Comment:
Same comment as above. Avoid doing twice the same type of lookup into `g2lMap`.
```suggestion
		final Set<Node> localSubjects = g2lMap.get( tp.asJenaTriple().getSubject() );
		if( localSubjects != null ) {
			final Iterator<Node> it = localSubjects.iterator();
```

Reply:
Also, as another small optimization, before even attempting to replace the subject of the given triple pattern, check that the subject is a URI. If it is not, put `tp` into `workingSet`.